Added debugging documentation for when verification is slow. (#1981)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

## 24.8. Verification

> In this section, we suggest a methodology to figure out [why a single assertion might not hold](#sec-verification-debugging), we propose methodologies to deal with [assertions that slow a proof down](#sec-verification-debugging-slow), we explain how to [verify assertions in parallel or in a focused way](#sec-assertion-batches) to devote more time to prove them, and we also give some more examples of [useful options and attributes to control verification](#sec-command-line-options-and-attributes-for-verification).

Reviewer: Why does verifying assertions in parallel devote more time to prove them?

Reviewer: methodologies → techniques
(A methodology (singular) is already a set of techniques, so methodologies are sets of techniques)

Author: >Why does verifying assertions in parallel devote more time to prove them?

It devotes more time if you sum all the times spent by each core used by the verifier. But I'm removing this remark as it's not very relevant.

> methodologies → techniques

Reworded. Note that I think that although a "methodology" is a "set of techniques", it's a subtype of it, because it feels like the techniques are homogeneous in some way. That's why I'll keep "methodology" for the first part, and "techniques" for the second.

  <br><br>`assert forall i | i == m :: P(m);` |  `assert P(m);`<br>`assert forall i | i == m :: P(i);`
  `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br><br><br>&nbsp;&nbsp;`  var x := m(a);`<br>&nbsp;&nbsp;`  assert P(x);` | `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A(k);`<br>&nbsp;&nbsp;`  assert forall x :: B(k, x) ==> P(x);`<br>&nbsp;&nbsp;`  var x := m(k);`<br>&nbsp;&nbsp;`  assert P(x);`

### 24.8.2. Verification debugging when verification is slow {#sec-verification-debugging-slow}

> In this section, we describe methodologies to apply in the case when verification is slower than expected, does not terminate, or timeouts.

Reviewer 2: How about a first subsection that suggests verifying with more, smaller assertion batches. Personally I would suggest first trying one batch per assertion before moving on to the techniques below. You can link the reader to https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-assertion-batches for more details.

#### 24.8.2.1. `assume false;` {#sec-assume-false}

> Assuming `false` is a powerful way to short-circuit the verifier and stop verification at a given point[^explainer-assume-false], and since the final compilation steps do not accept this command, it is safe to use it during development.

Reviewer 2: Just FYI, I'm a little wary of documenting this so strongly since its effectiveness is dependent on undefined behavior of the verifier: https://github.com/dafny-lang/dafny/issues/1441

This looks like great content that seems to work well now, just might need a caveat like what I discuss in that issue.

Reviewer: I'm bothered by the fact that the compiler rejects `assume false` — it seems quite arbitrary.  I'm not sure it's worth documenting here.

Author: > Just FYI, I'm a little wary of documenting this so strongly since its effectiveness is dependent on undefined behavior of the verifier: https://github.com/dafny-lang/dafny/issues/1441

Interesting. What about making "assume false; X" officially recognized by the Dafny verifier (like what you suggested `unverified { X }` so that it will not emit any assert in that block?
Since it seems this works well, I'll reword to ensure I describe it as "experimental" or "know-how".

> I'm bothered by the fact that the compiler rejects assume false — it seems quite arbitrary. I'm not sure it's worth documenting here.

I think it's for the same legitimate reason as rejecting functions or methods without bodies if they are not `{:extern}`. There is nothing to back an `assume` (not just `assume false;`, but every `assume`)

Author: Here is my new reformulation that is less emphasizing:

"Assuming `false` is an empirical way to short-circuit the verifier and usually stop verification at a given point"

Another similar command, `assert false;`, would also short-circuit the verifier, but it would still make the verifier try to prove `false`, which can also lead to timeouts.

[^explainer-assume-false]: `assume false` tells the Dafny verifier "Assume everything is true from this point of the program". The reason is that, 'false' proves anything. For example, `false ==> A` is always true because it is equivalent to `!false || A`, which reduces to `true || A`, which reduces to `true`.

> Thus, let us say a program of this shape takes forever to terminate.

Reviewer: verify, not terminate

```dafny
method NotTerminating() {
   assert X;
   if b {
     assert Y;
     assume false;
   } else {
     assume false;
     assert Z;
     assert P;
   }
   assert W;
}
```

Now, if verification is slow, it means that `assert Y;` is the problem.
If verification is fast, it means that the problem lies in the `else` branch.
One trick to ensure we measure the verification time of the `else` branch and not the then branch is to move the first `assume false;` to the top of the then branch, along with a comment indicating that we are short-circuiting it for now.
Then, we can move the second `assume false;` down and identify which of the two assertions makes the verifier to be slow.

```dafny
method NotTerminating() {
   assert X;
   if b {
     assume false; // Short-circuit because this branch is verified anyway
     assert Y;
   } else {
     assert Z;
     assume false;
     assert P;
   }
   assert W;
}
```

If verification is fast, which of the two assertions `assert Z;` or `assert P;` causes the slowdown?[^answer-slowdown]

[^answer-slowdown]: `assert P;`.

> We now hope you know enough of `assume false;` to locate an assertion that makes verification slow.

Reviewer: It's not always the case that a single assertion causes slowness, unfortunately :/  It can be multiple assertions together causing a loop.

Author: I modified to use the plural instead of the singular.

Nest, we will describe some other strategies at the assertion level to figure out what happens and perhaps fix it.

#### 24.8.2.2. `assert ... by {}` {#sec-verification-debugging-assert-by}

If an assertion `assert X;` is slow, it is possible that calling a lemma or invoking other assertions can help proving it: The postcondition of this lemma, or the added assertions, could help the Dafny verifier figure out faster how to prove the result.

```dafny

>   assert LEMMA_PRECONDITION;

Reviewer: This assert is unnecessary in most cases, since it's exactly the precondition of the following function.

Author: ```dafny
assert SOMETHING_HELPING_TO_PROVE_LEMMA_PRECONDITION;
```

  LEMMA();
  assert X;
...
lemma () 
  requires LEMMA_PRECONDITION
  ensures X { ... }
```

However, this approach has the problem that it exposes the asserted expressions and lemma postconditions not only for the assertion we want to prove faster,
but also for every assertion that appears afterwards. This can result in slowdowns[^verifier-lost].
A good practice consists of wrapping the intermediate verification steps in an `assert ... by {}`, like this:

```dafny
  assert X by {
    assert LEMMA_PRECONDITION;
    LEMMA();
  }
```

Now, only `X` is available for the Dafny Verifier to prove the rest of the method.

> [^verifier-lost]: By default, the expression of an assertion or a precondition is added to the knowledge base of the Dafny verifier for further assertions or postconditions. However, this is not always desirable, because if the verifier has too much knowledge, it might get lost trying to prove something in the wrong direction.

Reviewer: Further *within the same basic block*

#### 24.8.2.3. Labeling and revealing assertions {#sec-labeling-revealing-assertions}

Another way to prevent assertions or preconditions from cluttering the verifier[^verifier-lost] is to label and reveal them.
Labeling an assertion has the effect of "hiding" its result, until there is a "reveal" calling that label.

The example of the [previous section](#sec-verification-debugging-assert-by) could be written like this.

```dafny
  assert precond: LEMMA_PRECONDITION;
  assert X by {
    reveal precond;
    LEMMA();
  }
```

Similarly, if a precondition is only needed to prove a specific result in a method, one can label and reveal the precondition, like this:

```dafny
method Slow(i: int, j: int)
  requires greater: i > j {
  
  assert i >= j by {
    reveal greater;
  }
}
```

#### 24.8.2.4. Non-opaque `function method` {#sec-non-opaque-function-method}

Functions are normally used for specifications, but their functional syntax is sometimes also desirable to write application code.

> However, doing so naively results in the body of a `function method Fun()` be available for every caller, which can result in the verifier to timeout or get extremely slow[^verifier-lost].

Reviewer: `can cause the verifier to time out` or `can result in the verifier timing out`

```dafny
type byte = x | 0 <= x < 256
```

One of the problems of this approach is that additions, substractions and multiplications do not enforce the result to be in the same bounds, so it would have to be checked, and possibly truncated with modulos. For example:

```dafny
  var a: byte := 250;
  var b: byte := 200;
  var c := b - a;               // inferred to be an 'int', its value will be 50.
  var d := a + b;               // inferred to be an 'int', its value will be 450.
  var e := (a + b) % 256;       // still inferred to be an 'int'...
  var f: byte := (a + b) % 256; // OK
```

A better solution consists of creating a [newtype](#sec-newtypes) that will have the ability to check bounds of arithmetic expressions, and can actually be compiled to bitvectors as well.

```dafny
newtype {:nativeType "char"} byte = x | 0 <= x < 256
  var a: byte := 250;
  var b: byte := 200;
  var c := b - a; // OK, inferred to be a byte
  var d := a + b; // Error: cannot prove that the result of a + b is of type `byte`.
  var f := ((a as int + b as int) % 256) as byte // OK
```

> One might consider refactoring this code into specific functions if used over and over.

Reviewer: ```suggestion
One might consider refactoring this code into separate functions if used over and over.
```

#### 24.8.2.6. Nested loops {#sec-nested-loops}

In the case of nested loops, the verifier might timeout sometimes because of the information available[^verifier-lost].

> One way to mitigate this problem, when it happens, is to isolate the inner loop byt refactoring it to another method, with suitable pre and postconditions that will usually assume and prove the invariant again.

Reviewer: ```suggestion
One way to mitigate this problem, when it happens, is to isolate the inner loop by refactoring it into a separate method, with suitable pre and postconditions that will usually assume and prove the invariant again.
```

For example,

```dafny
`while X
   invariant Y
 {
   while X'
     invariant Y'
   {
 
   }
 }
```

could be refactored as this:

```dafny
`while X
   invariant Y
 {
   innerLoop();
 }
...
method innerLoop()
  require Y'
  ensures Y'
```

> In the nest section, when everything can be proven in a timely manner, we explain another strategy to decrease proof time by parallelizing it if needed, and making the verifier focused on certain parts.

Reviewer: ```suggestion
In the next section, when everything can be proven in a timely manner, we explain another strategy to decrease proof time by parallelizing it if needed, and making the verifier focus on certain parts.
```

[^answer-slowdown]: `assert P;`.

> We now hope you know enough of `assume false;` to locate assertions that makes verification slow.

Reviewer: ```suggestion
We now hope you know enough of `assume false;` to locate assertions that make verification slow.
```

> [^smaller-batches]: To create a smaller batch, Dafny duplicates the assertion batch, and transforms each assertion into an assumption into exactly one batch, so that assertions are verified only in one batch. This results in "easier" formulas for the verifier because it has less to prove, but it takes more overhead because every verification instance have a common set of axioms and there is no knowledge sharing between instances because they run independently.

Reviewer: "into an assumption into exactly one batch" this is not new but I don't understand it

Reviewer: "abitrarily" is odd; why is it arbitrary?

Author: It's because it's about splitting with heuristic costs, see here:
https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-vcs_max_cost

  "

"$DOTNET" "$BOOGIE" $BOOGIE_OPTIONS $PROVER_OPTIONS "$@"
#Uncomment if you want to use the executable instead of the DLL

> #"$BOOGIE" $BOOGIE_OPTIONS $PROVER_OPTIONS "$@"

Reviewer: ?

Author: The space triggered the automatic numbering to complain that 2 sections were missing. Removing this space made the [regex not create a false match](https://github.com/dafny-lang/dafny/blob/master/docs/DafnyRef/Numbers.java#L16) anymore:

Author: THe `\h+ you mean?`That would be even worse, my workaround would not work anymore.

Reviewer: No, the `h*` before the `#`

Reviewer: But it doesn't really matter, let's forget about this for now ^^
